build-tools/packages/build-cli/src/library/writeFileIfContentsDiffers.ts

```diff
@@ -0,0 +1,33 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { type PathLike } from "node:fs";
+import { readFile, writeFile } from "node:fs/promises";
+
+/**
+ * Writes to a file, but first reads the contents to check if it matches the desired content. If it does, the operation
+ * is skipped. Newlines are _always_ normalized to be line-feeds only for comparison. This effectively means that
+ * newline differences are ignored when comparing the contents. However, if the file is written, then it will be written
+ * as provided. For example, if `contents` is using CRLF newlines and the file is written, it will be written with CRLF
+ * newlines.
+ *
+ * @param filePath - The path to the file to write.
+ * @param contents - The contents to write to the file.
+ * @returns True if the file was written; false otherwise.
+ */
```

Comment:
Can we add a note about performance improvements?
This helper fully reads and closes a file before writing it (if needed).
In the common case that writing is needed, it will be more efficient to open the file for read+write to start and keep it open until any writing is dealt with.